client/lib/signup/cart.js

```diff
@@ -14,6 +14,26 @@ import productsListFactory from 'lib/products-list';
 const productsList = productsListFactory();
 import { cartItems, fillInAllCartItemAttributes } from 'lib/cart-values';
 
+function preprocessCartForServer( cart ) {
+	const newCart = { ...cart };
+
+	if (
+		! newCart.coupon &&
+		! newCart.is_coupon_applied &&
+		! newCart.is_coupon_removed &&
+		typeof document !== 'undefined'
+	) {
+		const coupon = url.parse( document.URL, true ).query.coupon;
+
+		if ( coupon ) {
+			newCart.coupon = coupon;
+			newCart.is_coupon_applied = false;
+		}
+	}
+
+	return newCart;
+}
```

Comment:
should this and the one above be combined/composed? seems like duplicate logic that's asking to get out of sync

Reply:
Yes, definitely. Thanks for catching this. I moved this function to `lib/cart-values` so there's no need for the duplicate logic. I'm importing the function as needed now.